### ISSUE TITLE
Type information is somehow lost

### ISSUE THREAD

**Issue by sorawee** [brownplt/pyret-lang#1311]
```
data Item: item end
map(lam(x): num-to-string(x) end, [list: item])
```

reports an error:

```
The type constraint Number was incompatible with the type constraint Item
```

So the type checker definitely knows that `x` is an `Item`.

However:

```
data Item: item end

map(lam(x) block:
    cases (Item) x:
      | item => 1
    end
  end, [list: item])
```

says it is unable to infer the type of `x`

**Comment by justinpombrio** [comment]
Here's another instance of the same issue:

    fun foo(x :: List<List<Number>>):
      for each(y from x):
        y.get(0) + 1
      end
    end

**Comment by sorawee** [comment]
This is running with the typechecker (instead of "Run", use "Type-check and run (beta)")

**Comment by SethPoulsen** [comment]
Oops. I just noticed that myself. Thanks.

**Comment by blerner** [comment]
This looks like a dupe of #1309.

**Comment by sorawee** [comment]
Yes and no. I meant #1309 to focus on "there is an additional information, and the type checker should use it". In this issue, however, I am focusing on, "why does it lose information in the first place?".
